source/common/quic/http_datagram_handler.h

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include "envoy/http/codec.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/common/logger.h"
+#include "source/common/http/header_map_impl.h"
+
+#include "absl/strings/string_view.h"
+#include "quiche/common/simple_buffer_allocator.h"
+#include "quiche/quic/core/http/quic_spdy_stream.h"
+
+namespace Envoy {
+namespace Quic {
+
+// Enables HTTP Datagrams and the Capsule Protocol support based on RFC 9297. This class is used
+// both on the receive side (HTTP/3 codec -> Envoy) and the send side (Envoy -> HTTP/3 codec) of
+// the QUIC client and server streams in Envoy.
+class HttpDatagramHandler : public quic::QuicSpdyStream::Http3DatagramVisitor,
+                            public quiche::CapsuleParser::Visitor,
+                            protected Logger::Loggable<Logger::Id::quic_stream> {
+public:
+  // Does not take ownership of the QuicSpdyStream. "stream" must refer to a valid object
+  // that outlives HttpDatagramHandler.
+  explicit HttpDatagramHandler(quic::QuicSpdyStream* stream);
```

Comment:
nit: Consider making this a ref instead of a pointer. QUICHE tends to use pointers when passing in non-const references, but Envoy uses non-const refs.

Reply:
Very good point! Done.